artificial_algorand_contract/escrow/escrow_contract.py

```diff
@@ -21,218 +27,246 @@
+    )  # global_ints_scheme
+
+    on_opt_in = Seq(
+        App.localPut(Gtxn[0].sender(), Bytes(ASSET_NAME), Int(0)),
+        App.localPut(Gtxn[0].sender(), Bytes(STABLE_NAME), Int(0)),
+        App.localPut(Gtxn[0].sender(), Bytes("last_msg"), Bytes("OptIn OK.")),
+        SucceedSeq,
+    )  # always allow user to opt in
+
+    on_close_out = FailWithMsg("Only allow 0 $ART$ user, or lose $ART$.")
+
+    on_update_app = Return(
+        And(
+            Global.creator_address() == Gtxn[0].sender(),  # TODO: manager
+            Ed25519Verify(
+                data=Gtxn[0].application_args[0],
+                sig=Gtxn[0].application_args[1],
+                key=Gtxn[0].application_args[2],
+            ),  # security, should use a password, high cost is ok.
+        )
+    )
+
+    on_deleteapp = on_update_app  # Same security level as update_app
 
-    handle_no_op = Cond(
+    on_call = Cond(
         [
             And(
-                Global.group_size() == Int(1),
-                Txn.application_args[0] == Bytes("escrow"),
+                Gtxn[0].application_args[0] == Bytes("mint"),
             ),
-            escrow,
+            on_mint,
         ],
         [
             And(
-                Global.group_size() == Int(1),
-                Txn.application_args[0] == Bytes("redeem"),
-                App.localGet(Txn.sender(), Bytes(STABLE_NAME)) >= Txn.asset_amount(),
-                # TODO: make sure correct logic
+                Gtxn[0].application_args[0] == Bytes("burn"),
+                App.localGet(Gtxn[1].asset_sender(), Bytes(STABLE_NAME))
+                >= Gtxn[1].asset_amount(),
+                # TODO:discuss: move to on_burn? needed? (TEAL has integer underflow)
+                # correct logic depend on ACID (atomicity, consistency, isolation, durability).
+                # cannot be used to cheat (will not parallel) for ACID.
             ),
-            redeem,
+            on_burn,
         ],
         [
             Int(1),
-            Return(Int(0)),  # Fail if no correct args.
+            FailWithMsg("wrong args"),
         ],
     )
 
     program = Cond(
-        [Txn.application_id() == Int(0), handle_creation],
-        [Txn.on_completion() == OnComplete.OptIn, handle_opt_in],
-        [Txn.on_completion() == OnComplete.CloseOut, handle_close_out],
-        [Txn.on_completion() == OnComplete.UpdateApplication, handle_update_app],
-        [Txn.on_completion() == OnComplete.DeleteApplication, handle_deleteapp],
-        [Txn.on_completion() == OnComplete.NoOp, handle_no_op],
+        [Gtxn[0].application_id() == Int(0), on_creation],
+        [Gtxn[0].on_completion() == OnComplete.NoOp, on_call],
+        [Gtxn[0].on_completion() == OnComplete.CloseOut, on_close_out],
+        [Gtxn[0].on_completion() == OnComplete.UpdateApplication, on_update_app],
+        [Gtxn[0].on_completion() == OnComplete.DeleteApplication, on_deleteapp],
+        [Gtxn[0].on_completion() == OnComplete.OptIn, on_opt_in],
```

Comment:
Question: 
1) can Gtxn[1] Gtxn[2] ... also be CallApp type transaction
2) if 1) is true, can you specifiy another total 2048-byte arguments for each Gtxn
3) can you group more than 16 txns in TEAL version 6?

Reply:
1. Yes, they can . But the `Gtxn[0]` has to be the CallApp type (OnComplete.NoOp). I remembered seeing that the latter transaction cans see the new states of the first ones, but can't find the docs.
2. Yes.
3. No, you can't. see [Algorand parameter tables](https://developer.algorand.org/docs/get-details/parameter_tables/#others)